File: backend/services/clientService.js
const { ClientCreationError, ClientNotFoundError, ClientUpdateError, ClientDeleteError } = require('../middleware/customErrors');
const Client = require('../models/ClientModels');

const clientService = {
    
    // Create a new client
    async createClient(ownerId, clientData) {
        try {
            if (!ownerId || !clientData) {
                throw new Error("Missing userId or client Data");
              }
            clientData.owner = ownerId; // Ensure the client is associated with the user
            const client = new Client(clientData);
            const savedClient = await client.save();
            return savedClient;
        } catch (error) {
            console.error("Error while creating client: ", error);
            throw new ClientCreationError();
        }
    },

    // Retrieve a client by its ID and owner ID
    async getClientById(ownerId, clientId) {
        try {
            const client = await Client.findById(clientId);
            if (client.owner.toString() !== ownerId) {
                throw new ClientNotFoundError();
            }
            return client;
        } catch (error) {
            throw new ClientNotFoundError();
        }
    },

    // Retrieve all clients for a specific user
    async getClientsByOwnerId(ownerId) {
        try {
            return await Client.find({ owner: ownerId });
        } catch (error) {
            throw new ClientNotFoundError();
        }
    },

    // Update a client by its ID and owner ID
    async updateClientById(ownerId, clientId, updateData) {
        try {
            const client = await Client.findById(clientId);
            if (client.owner.toString() !== ownerId) {
                throw new ClientNotFoundError();
            }
            return await Client.findByIdAndUpdate(clientId, updateData, { new: true });
        } catch (error) {
            throw new ClientUpdateError();
        }
    },

    // Delete a client by its ID and owner ID
    async deleteClientById(ownerId, clientId) {
        try {
            const client = await Client.findById(clientId);
            if (client.owner.toString() !== ownerId) {
                throw new ClientNotFoundError();
            }
            return await Client.findByIdAndDelete(clientId);
        } catch (error) {
            throw new ClientDeleteError();
        }
    }

};

module.exports = clientService;

File: frontend/lib/services/client_service.dart
import 'package:http/http.dart' as http;
import '../models/user.dart';
import './request_service.dart'; // import the AuthenticatedService class

class ClientService {
  final String baseUrl;
  final User? user;
  final AuthenticatedService authService; // Use AuthenticatedService here

  ClientService({required this.baseUrl, this.user})
      : authService = AuthenticatedService(baseUrl: baseUrl, user: user);

  Future<http.Response> createClient(Map<String, dynamic> data) async {
    return authService.postRequest('/api/clients/', data);
  }

  Future<http.Response> getClientById(String clientId) async {
    return authService.getRequest('/api/clients/$clientId');
  }

  Future<http.Response> getClientsByOwnerId(String ownerId) async {
    return authService.getRequest('/api/clients/owner/$ownerId');
  }

  Future<http.Response> updateClientById(
      String clientId, Map<String, dynamic> data) async {
    return authService.putRequest('/api/clients/$clientId', data);
  }

  Future<http.Response> deleteClientById(String clientId) async {
    return authService.deleteRequest('/api/clients/$clientId');
  }
}
